### ISSUE TITLE
Caller stacktrace is missing on RuntimeExceptions thrown from ExceptionUtil.rethrow 

### ISSUE THREAD

**Issue by sancar** [hazelcast/hazelcast#17202]
After the fix https://github.com/hazelcast/hazelcast/pull/15441 we have removed exception stack trace rewrite on the server. 
Since the client proxies do not use correct method (which seems to be `invocationFuture.joinInternal`), the exceptions that come from the remote are thrown with the stack trace on the remote. 

This makes it hard to follow the stack trace. ~~The solution is already available on the member side. This issue is to address the problems on the client-side.~~ See next comment.

**Comment by sancar** [comment]
After some investigation, I have realized that this is a problem for the core side as well. 
We have lots of places that we use the following pattern 
```
try {
     invocationFuture.get()
}catch(Exception e){
    throw retrhow(e);
}
```

If the exception is RuntimeException, it is not wrapped and thrown directly. In that case, we can not follow where the exception is thrown from. Here is an example of a put from a member.

```
java.lang.RuntimeException: here comes the bug
	at com.hazelcast.map.impl.operation.PutOperation.runInternal(PutOperation.java:36)
	at com.hazelcast.map.impl.operation.MapOperation.run(MapOperation.java:112)
	at com.hazelcast.spi.impl.operationservice.Operation.call(Operation.java:184)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:227)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:216)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:422)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:166)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:136)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.executeRun(OperationThread.java:123)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)
```

Same problem on the client side:

```
java.lang.RuntimeException: here comes the bug
	at com.hazelcast.map.impl.operation.PutOperation.runInternal(PutOperation.java:36)
	at com.hazelcast.map.impl.operation.MapOperation.run(MapOperation.java:112)
	at com.hazelcast.spi.impl.operationservice.Operation.call(Operation.java:184)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:227)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:216)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.run(OperationExecutorImpl.java:411)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationExecutorImpl.runOrExecute(OperationExecutorImpl.java:438)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvokeLocal(Invocation.java:597)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.doInvoke(Invocation.java:582)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke0(Invocation.java:541)
	at com.hazelcast.spi.impl.operationservice.impl.Invocation.invoke(Invocation.java:238)
	at com.hazelcast.spi.impl.operationservice.impl.InvocationBuilderImpl.invoke(InvocationBuilderImpl.java:59)
	at com.hazelcast.client.impl.protocol.task.AbstractPartitionMessageTask.processInternal(AbstractPartitionMessageTask.java:51)
	at com.hazelcast.client.impl.protocol.task.AbstractAsyncMessageTask.processMessage(AbstractAsyncMessageTask.java:71)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.initializeAndProcessMessage(AbstractMessageTask.java:153)
	at com.hazelcast.client.impl.protocol.task.AbstractMessageTask.run(AbstractMessageTask.java:116)
	at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:180)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:172)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:140)
	at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.executeRun(OperationThread.java:123)
	at com.hazelcast.internal.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:102)
```

These exceptions should contain the stack trace of the caller as well. But they only have the remote stack trace.

**Comment by mmedenjak** [comment]
Reopening as the change was reverted.

**Comment by sancar** [comment]
Leaving here as a note. The following PR's are also reverted. 
https://github.com/hazelcast/hazelcast/pull/17438
https://github.com/hazelcast/hazelcast/pull/17437
These need to be handled in the same pr that fixes this issue.

**Comment by sancar** [comment]
We did not see changing which layer the cause can be found as a problem. Now, it is clear that it was not a correct decision, because we are using the cause to decide the logic internally(our users also can use this info). And hiding the cause a couple of layers deep broke that logic in some places.
See https://github.com/hazelcast/hazelcast/issues/17506. 

So at this point, I have to revert the pr(https://github.com/hazelcast/hazelcast/pull/17212)
so that in our internals we will not mess with the cause hierarchy. In proxies, we can use AbstractInvocationFuture.joinInternal instead to have the caller stack trace info. There are a couple of downsides:

We will still mess up with the cause hierarchy for the hazelcast users. It will not be possible to write code that depends on the cause of an exception thrown by hazelcast proxies.
See this test as an example. https://github.com/hazelcast/hazelcast/pull/17212/files#diff-50e25ec5ff983e944a9621d48bcb5103R54
The original thrown exception was HazelcastException caused by ClassNotFoundException.
When we use joinInternal, it will be HazelcastException caused by HazelcastException caused by ClassNotFoundException.
There are lots of usages of rethrow/catch. It might not be possible to find and replace all places at once. But if we miss any, it will not be broken. We will be missing caller stack trace info in the same cases.
We need to be aware that we should use joinInternal when call is from user thread. If a new code uses rethrow/catch again, it will again be lacking the caller stack trace info.
Given that we are breaking the cause class hierarchy for the users, we might choose to

remove tryWrapInSameClass even from joinInternal. This means that we will not have any clue about the caller the stacktrace.
OR
Put back exception stacktrace rewrite and solve the race problem by cloning the exception. We have removed this in 4.0 for the sake of making exceptions immutable. The pr that we made the change. #15441
Related issue. #6247
